crates/storage/provider/src/change.rs

```diff
@@ -0,0 +1,1421 @@
+        let mut storage_changeset_cursor = tx.cursor_dup_write::<tables::StorageChangeSet>()?;
+        for (block_index, mut storage_changes) in self.0.storage.into_iter().enumerate() {
+            let block_number = first_block + block_index as BlockNumber;
+
+            tracing::trace!(target: "provider::reverts", block_number=block_number,"Writing block change");
+            // sort changes by address.
+            storage_changes.par_sort_unstable_by_key(|a| a.address);
+            for PlainStorageRevert { address, wiped, storage_revert } in storage_changes.into_iter()
+            {
+                let storage_id = BlockNumberAddress((block_number, address));
+
+                let mut storage = storage_revert
+                    .into_iter()
+                    .map(|(k, v)| (H256(k.to_be_bytes()), v))
+                    .collect::<Vec<_>>();
+                // sort storage slots by key.
+                storage.par_sort_unstable_by_key(|a| a.0);
+                tracing::trace!(target: "provider::reverts","Writting revert for {:?}", address);
+                // If we are writing the primary storage wipe transition, the pre-existing plain
+                // storage state has to be taken from the database and written to storage history.
+                // See [StorageWipe::Primary] for more details.
+                let mut wiped_storage = Vec::new();
+                if wiped {
+                    tracing::trace!(target: "provider::reverts", "wipe storage storage changes");
+                    if let Some((_, entry)) = storages_cursor.seek_exact(address)? {
+                        wiped_storage.push((entry.key, entry.value));
+                        while let Some(entry) = storages_cursor.next_dup_val()? {
+                            wiped_storage.push((entry.key, entry.value))
+                        }
+                    }
+                }
+                tracing::trace!(target: "provider::reverts", "storage changes: {:?}",storage);
+                // if empty just write storage reverts.
+                if wiped_storage.is_empty() {
+                    for (key, old_value) in storage {
+                        storage_changeset_cursor.append_dup(
+                            storage_id,
+                            StorageEntry { key, value: old_value.to_previous_value() },
+                        )?;
+                    }
+                } else {
+                    // if there is some of wiped storage, they are both sorted, intersect both of
+                    // them and in conflict use change from revert (discard values from wiped
+                    // storage).
+                    let mut wiped_iter = wiped_storage.into_iter();
+                    let mut revert_iter = storage.into_iter();
+
+                    // items to apply. both iterators are sorted.
+                    let mut wiped_item = wiped_iter.next();
+                    let mut revert_item = revert_iter.next();
```

Comment:
this could be simplified with a custom iterator impl

Reply:
https://github.com/paradigmxyz/reth/pull/4595